### PR TITLE
Enable fts for mirrorless and log doublefault in gp_configuration_history

### PR DESCRIPTION
**Pull request by divyeshddv**
Update ftsprobe process to log each double fault occurence into
gp_configuration_history with the message

Also keep track of contents that are currently under doublefault.
Whenever any content comes out of double fault, update that status into
gp_configuration_history

gp_configuration_history log message:
```
 2023-12-13 09:57:54.000504-05 |    5 | FTS: double fault detected for content id 0
 2023-12-13 09:59:10.074589-05 |    5 | FTS: segment started for content id 0
```

Enable fts for mirrorless cluster, and also log segments going down/up into gp_configuration_history
```
 2023-12-13 09:57:54.000504-05 | 2    | FTS: content id 0 dbid 2 is down.  
 2023-12-13 09:59:10.074589-05 | 2    | FTS: content id 0 dbid 2 is now up 
 ```

### CODE REVIEW

File: src/test/isolation2/expected/fts_doublefault.out [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+(1 row)
+-1U: select pg_ctl_start(datadir, port) from gp_segment_configuration where role = 'p' and content = 1;
+ pg_ctl_start                                     
+--------------------------------------------------
+ waiting for server to start done
+server started
+ 
+(1 row)
+
+select dbid, description from gp_configuration_history where time > (select time from last_timestamp);
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/test/isolation2/expected/fts_doublefault.out:114]: Lets also do this query after every event (double fault OR coming out of double fault). Just easier to read the test and follow along.

File: src/test/isolation2/expected/fts_doublefault.out [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ dbid | description                                                                   
+------+-------------------------------------------------------------------------------
+ 2    | FTS: update role, status, and mode for dbid 2 with contentid 0 to m, d, and n 
+ 5    | FTS: update role, status, and mode for dbid 5 with contentid 0 to p, u, and n
```
Comment (soumyadeep2007) [inline src/test/isolation2/expected/fts_doublefault.out:118]: How come we didn't get `FTS: segment started for content id` in the messages?

File: src/test/isolation2/expected/fts_doublefault.out [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+-- stop primary in order to trigger double fault for content 1
+select pg_ctl((select datadir from gp_segment_configuration c where c.role='p' and c.content=1), 'stop');
+ pg_ctl 
+--------
+ OK     
+(1 row)
+
+-- trigger double fault on content 1 (FTS_PROMOTE_FAILED)
+select gp_request_fts_probe_scan();
```
Comment (soumyadeep2007) [inline src/test/isolation2/expected/fts_doublefault.out:93]: Lets do another probe while we are in double fault to test the de-duplication (we shouldn't be emitting 2 double fault messages for the same content id).

File: src/backend/fts/ftsprobe.c [hunk 4]
```diff
@@ -1017,6 +1017,22 @@ processResponse(fts_context *context)
 		{
 			case FTS_PROBE_SUCCESS:
 				Assert(IsPrimaryAlive);
+				if (list_length(*doublefaultdetected) > 0)
+				{
+					ListCell *cell;
+					foreach(cell, *doublefaultdetected)
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/backend/fts/ftsprobe.c]: Done

File: src/include/postmaster/ftsprobe.h [hunk 5]
```diff
@@ -112,7 +112,8 @@ typedef struct
 {
 	int num_pairs; /* number of primary-mirror pairs FTS wants to probe */
 	fts_segment_info *perSegInfos;
+	List *doublefaultdetected;
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/include/postmaster/ftsprobe.h:115]: ftscontext stays for one single fts probe. 
We create a new ftscontext one on every probe. 
We need the doublefault info throughout the loop, across multiple probes. I think we might have to stick with double pointers for that.

File: src/backend/fts/fts.c [hunk 6]
```diff
@@ -267,13 +267,59 @@ probeWalRepUpdateConfig(int16 dbid, int16 segindex, char role,
 	}
 }
 
+void
+probeUpdateConfigHist(int16 dbid, int16 segindex, bool IsSegmentAlive)
+{
+	/*
+	 * Insert new tuple into gp_configuration_history catalog.
+	 */
+	{
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/backend/fts/fts.c]: Done! thanks, must have missed on refactoring

File: src/test/isolation2/expected/fts_doublefault.out (hunk 2 above)
Comment (divyeshddv) [reply src/test/isolation2/expected/fts_doublefault.out:129]: Had no ftsprobe after getting the segments up. Fixed

File: src/test/isolation2/expected/fts_doublefault.out (hunk 1 above)
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/test/isolation2/expected/fts_doublefault.out]: Querying while doublefault seems to not work, getting error on psql

File: src/test/isolation2/expected/fts_doublefault.out (hunk 3 above)
Comment (divyeshddv) [reply src/test/isolation2/expected/fts_doublefault.out:99]: Done

File: src/backend/fts/ftsprobe.c [hunk 7]
```diff
@@ -1017,6 +1017,22 @@ processResponse(fts_context *context)
...
+					{
+						if (lfirst_int(cell) == primary->config->dbid)
+						{
+							StartTransactionCommand();
+							GetTransactionSnapshot();
+							probeUpdateConfigHist(primary->config->dbid, primary->config->segindex, true);
+							CommitTransactionCommand();
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/backend/fts/ftsprobe.c]: updateDoubleFaultStateConfigHist is in fts.c which doesn't import Transaction related functions. Created a separate function in ftsprobe.c to do the same

File: src/test/isolation2/expected/fts_doublefault.out (hunk 3 above)
Comment (soumyadeep2007) [reply src/test/isolation2/expected/fts_doublefault.out:116]: Let's remove all these tests for content = 1. Content = 0 should suffice and we can do the de-duplication test I mentioned above for content = 0 itself.
This will help cut down on test runtime quite a lot, and will help readability.

PS: 
1. When you trigger another probe to test the de-duplication of the "double fault detected" message, add a comment: "Subsequent probes should not create duplicate "double fault detected" messages.
2. Let's also test the de-duplication of "out of double fault" messages, and add a similar comment as 1.

File: src/test/isolation2/expected/fts_doublefault.out [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+(exited with code 0)
+
+-- loop while segments come in sync
+select wait_until_all_segments_synchronized();
+ wait_until_all_segments_synchronized 
+--------------------------------------
+ OK                                   
+(1 row)
+
+-1U: select dbid, description from gp_configuration_history where time > (select time from last_timestamp) order by time;
```
Comment (soumyadeep2007) [inline src/test/isolation2/expected/fts_doublefault.out:197]: This isn't really necessary for the main goal of this test?

File: src/test/isolation2/expected/fts_doublefault.out [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
```
Comment (soumyadeep2007) [inline src/test/isolation2/expected/fts_doublefault.out:227]: This isn't really necessary for the main goal of this test?

File: src/test/isolation2/expected/fts_doublefault.out [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+ 6    | FTS: update role, status, and mode for dbid 6 with contentid 1 to m, u, and s 
+ 5    | FTS: update role, status, and mode for dbid 5 with contentid 0 to p, u, and s 
+ 2    | FTS: update role, status, and mode for dbid 2 with contentid 0 to m, u, and s 
+ 5    | FTS: update role, status, and mode for dbid 5 with contentid 0 to m, d, and n 
+ 2    | FTS: update role, status, and mode for dbid 2 with contentid 0 to p, u, and n 
+ 2    | FTS: update role, status, and mode for dbid 2 with contentid 0 to p, u, and s 
+ 5    | FTS: update role, status, and mode for dbid 5 with contentid 0 to m, u, and s 
+(16 rows)
+
+-- verify no segment is down after recovery
```
Comment (soumyadeep2007) [inline src/test/isolation2/expected/fts_doublefault.out:248]: Don't think that this is necessary.

**Review by soumyadeep2007 — COMMENTED** [review]
Some more feedback on cutting down on test runtime and some cosmetic suggestions for the code.

Since we are storing the double fault state in memory, let me ask the following question:
What would happen if a double fault is detected for content = 0 and afterwards, the ftsprobe process crashes/terminates and restarts? The in-memory state is lost.
Would we be able to re-build the in-memory state correctly? I think we would, but it would lead to a duplicate "double fault" message in gp_configuration_history. This could be tested manually and in the test (using pg_terminate_backend on the ftsprobe process, triggering an FTS probe). Let's verify this manually first and then automate it.

File: src/test/isolation2/expected/fts_doublefault.out (hunk 3 above)
Comment (divyeshddv) [reply src/test/isolation2/expected/fts_doublefault.out:116]: Content 1 double fault is for FTS_PROMOTE_FAILED state while content 0 test is for FTS_PROBE_FAILED state. Are you suggesting removing the test for FTS_PROMOTE_FAILED? or have it done sequentially on content 0?

Comment (soumyadeep2007) [reply src/test/isolation2/expected/fts_doublefault.out:116]: Oh I completely missed the FTS_PROBE_FAILED vs FTS_PROMOTE_FAILED aspects. Never mind then.

File: src/backend/fts/ftsprobe.c [hunk 11]
```diff
@@ -971,14 +971,35 @@ updateConfiguration(CdbComponentDatabaseInfo *primary,
...
+ * segment pair goes into double fault or comes out of double fault.
+ * Since FTS probe doesn't have its own transaction, we need to create one
+ * to update the catalog and commit it once it is done.
+ */
+static void
+CreateTransactionUpdateConfigHist(CdbComponentDatabaseInfo *primary, Bitmapset **doubleFaultContentIds, bool isSegmentAlive)
+{
+	StartTransactionCommand();
+	GetTransactionSnapshot();
+	updateDoubleFaultStateConfigHist(primary->config->dbid,
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/backend/fts/ftsprobe.c]: Done

File: src/backend/fts/ftsprobe.c [hunk 12]
```diff
@@ -984,7 +987,9 @@ updateConfiguration(CdbComponentDatabaseInfo *primary,
  * to update the catalog and commit it once it is done.
  */
 static void
-updateDoubleFaultStatus(CdbComponentDatabaseInfo *primary, bool isSegmentAlive)
+updateDoubleFaultStatus(CdbComponentDatabaseInfo *primary,
+						bool isSegmentAlive,
+						bool hasMirrors)
```
Comment (huansong) [inline src/backend/fts/ftsprobe.c:992]: Mirrorless cluster doesn't have "double fault" so it's kinda confusing. Maybe name it `updateSegmentDownStatus` instead?

File: src/backend/fts/ftsprobe.c [hunk 13]
```diff
@@ -901,8 +907,11 @@ ftsResponseReady(fts_segment_info *ftsInfo)
 static bool
 updateConfiguration(CdbComponentDatabaseInfo *primary,
 					CdbComponentDatabaseInfo *mirror,
-					char newPrimaryRole, char newMirrorRole,
-					bool IsInSync, bool IsPrimaryAlive, bool IsMirrorAlive)
+					char newPrimaryRole,
+					char newMirrorRole,
+					bool IsInSync,
+					bool IsPrimaryAlive,
+					bool IsMirrorAlive)
```
Comment (huansong) [inline src/backend/fts/ftsprobe.c:914]: nit: unnecessary diff

File: src/test/isolation2/expected/fts_doublefault.out [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+ gp_request_fts_probe_scan 
+---------------------------
+ t                         
+(1 row)
+-1U: select dbid, description from gp_configuration_history where time > (select time from last_timestamp) order by time;
+ dbid | description                                                                   
+------+-------------------------------------------------------------------------------
+ 2    | FTS: update role, status, and mode for dbid 2 with contentid 0 to m, d, and n 
+ 5    | FTS: update role, status, and mode for dbid 5 with contentid 0 to p, u, and n 
+ 5    | FTS: double fault detected for content id 0
```
Comment (huansong) [inline src/test/isolation2/expected/fts_doublefault.out:73]: Hmm I wonder why currently we don't log a gp_configuration_history item in double fault? For example, why didn't we have item below in the test case here:
```
FTS: update role, status, and mode for dbid 5 with contentid 0 to p, d, and n 
```
Is it because we won't update gp_segment_configuration in double fault? If so, why not?

This question is related to the PR in that I wondered why we are not doing the new updates in `updateConfiguration` so that: (1) we can save opening/closing transaction blocks; (2) it's easier to reason about - gp_configuration_history should only reflect changes done to gp_segment_configuration.

File: src/backend/fts/ftsprobe.c [hunk 15]
```diff
@@ -1241,21 +1353,34 @@ FtsWalRepInitProbeContext(CdbComponentDatabases *cdbs, fts_context *context)
 
 	int fts_index = 0;
 	int cdb_index = 0;
+	context->has_mirrors = false;
```
Comment (huansong) [inline src/backend/fts/ftsprobe.c:1356]: nit: no need this line as we will be relying on the return value of `gp_segment_config_has_mirrors()` anyway.

File: src/test/isolation2/expected/fts_doublefault.out (hunk 14 above)
Comment (divyeshddv) [reply src/test/isolation2/expected/fts_doublefault.out:73]: > Is it because we won't update gp_segment_configuration in double fault? If so, why not?
That is correct, we don't update gp_segment_configuration in case of double fault. 

For cluster with mirrors, when a double fault occurs, we want to make sure that we fix/start the segment which was acting primary at the time so that we don't lose out on any data. The acting mirror needs then to be recovered using gprecoverseg.
I believe that is why we want to maintain segment configuration before double fault, (or why we don't update it in gp_seg_conf)

File: src/backend/fts/ftsprobe.c [hunk 16]
```diff
@@ -1078,12 +1159,16 @@ processResponse(fts_context *context)
...
-						GP_SEGMENT_CONFIGURATION_ROLE_MIRROR,
-						IsInSync, IsPrimaryAlive, IsMirrorAlive);
+					if (context->has_mirrors)
+					{
+						is_updated |= updateConfiguration(
+							primary, mirror,
+							GP_SEGMENT_CONFIGURATION_ROLE_PRIMARY,
+							GP_SEGMENT_CONFIGURATION_ROLE_MIRROR,
+							IsInSync, IsPrimaryAlive, IsMirrorAlive);
+					}
```
Comment (huansong) [inline src/backend/fts/ftsprobe.c:1169]: As mentioned in previous comment, if we add a if check like this:
```
if (context->has_mirrors)
{
    elogif(..., "FTS skipping mirror down update for (content=%d) as mirrorless" ...);
}
else if (ftsInfo->result.isSyncRepEnabled && !IsMirrorAlive)
...
```
then we don't need the changes here.

File: src/backend/fts/ftsprobe.c [hunk 17]
```diff
@@ -1153,13 +1238,26 @@ processResponse(fts_context *context)
...
+						elog(WARNING, "ERROR: FTS double fault detected (content=%d) "
+								  "primary dbid=%d, mirror dbid=%d",
+								  primary->config->segindex, primary->config->dbid, mirror->config->dbid);
+					 }
+					else
+					{
+						elog(WARNING, "ERROR: FTS detected segment down (content=%d) "
+									  "primary dbid=%d",
+							 primary->config->segindex, primary->config->dbid);
+					}
```
Comment (huansong) [inline src/backend/fts/ftsprobe.c:1252]: This should be after the comment `Only log here, will handle it later...`

**Review by huansong — COMMENTED** [review]
Mostly good but let's sort out some of the logic in processing response for mirrorless better. Also don't forget to fly a dev pipeline to verify the changes.

File: src/backend/fts/ftsprobe.c [hunk 18]
```diff
@@ -1017,6 +1090,11 @@ processResponse(fts_context *context)
 		{
 			case FTS_PROBE_SUCCESS:
 				Assert(IsPrimaryAlive);
+
+				if (bms_is_member(primary->config->segindex, doubleFaultContentIds))
+					updateSegmentDownStatus(primary,
+											true, context->has_mirrors);
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/backend/fts/ftsprobe.c]: Yes, looks cleaner now

File: src/backend/fts/ftsprobe.c [hunk 19]
```diff
@@ -1025,16 +1103,19 @@ processResponse(fts_context *context)
...
+								primary, mirror,
+								GP_SEGMENT_CONFIGURATION_ROLE_PRIMARY,
+								GP_SEGMENT_CONFIGURATION_ROLE_MIRROR,
+								IsInSync, IsPrimaryAlive, IsMirrorAlive);
+							/*
+							 * If mirror was marked up in configuration, it must have
+							 * been marked down by updateConfiguration().
+							 */
+							AssertImply(SEGMENT_IS_ALIVE(mirror), is_updated);
+						}
```
[the remark this answers is not in the document]
Comment (divyeshddv) [reply src/backend/fts/ftsprobe.c]: We don't come down here for mirrorless, refactored the branch
